recipes/mipro/mipro_nb.py

```diff
@@ -325,88 +340,147 @@
+    """
+    content: List[str] = []
+    tool_calls: List[Dict[str, Any]] = []
+
+    if base_function.type == "json":
+        return {"role": "assistant", "content": output["raw"]}
+    elif base_function.type == "chat":
+        for content_block in output:
+            if content_block["type"] == "text":
+                content.append({"type": "text", "text": content_block["text"]})
+            elif content_block["type"] == "thought":
+                content.append(
+                    {"type": "text", "text": f"<think>{content_block['value']}</think>"}
+                )
+            elif content_block["type"] == "tool_call":
+                tool_calls.append(
+                    {
+                        "function": {
+                            "arguments": json.dumps(content_block["arguments"]),
+                            "name": content_block["name"],
+                        },
+                        "id": content_block["id"],
+                        "type": "function",
+                    }
+                )
             else:
-                raise ValueError(
-                    f"Output {block} must be a text, tool_call, or tool_result block."
+                warnings.warn(
+                    f"We do not support content block type: {content_block['type']}, dropping example.",
+                    UserWarning,
                 )
-            return "\n".join(rendered_output)
-    elif base_function.type == "json":
-        return output["raw"]
+                return None
     else:
         raise ValueError(f"Unsupported function type: {base_function.type}")
 
+    # Once we finish collecting all blocks, create one user message if there's any user content
+    output_message: Dict[str, Any] = {"role": "assistant"}
+    if content:
+        output_message["content"] = content
+    if tool_calls:
+        output_message["tool_calls"] = tool_calls
 
-def format_system_args(sample):
+    return output_message
+
+
+def sample_to_openai_messages(sample) -> List[Dict[str, Any]]:
     function_input = json.loads(sample["input"])
-    if "system" in function_input:
-        return function_input["system"]
-    else:
-        return ""
 
+    rendered_messages = []
+
+    # Add the system message to the rendered messages
+    # If there is data passed in or a system template there must be a system message
+    system = function_input.get("system", {})
+    if len(system) > 0 or base_variant.system_template:
+        if base_variant.system_template:
+            system_message = env.render_template("system", **system)
+            rendered_messages.append({"role": "system", "content": system_message})
+        else:
+            rendered_messages.append({"role": "system", "content": system})
+
+    # Add the input messages to the rendered messages
+    for message in function_input["messages"]:
+        rendered_message = render_message(message)
+        if rendered_message is None:
+            return None
```

Comment:
in all the places where we do this can you add a comment saying "Drop the entire example if any message is dropped"?